### PR TITLE
Add Ascend NPU support for SDXL fine-tuning and fix the model saving bug when using DeepSpeed.

### PR DESCRIPTION
**Pull request by HelloWorldBeginner**
# What does this PR do?

Added support for SDXL finetune on AscendNPU and fixed the bug causing the hang out when saving models using the deepspeed distributed framework. DeepSpeed requires saving weights on every device; saving weights only on the main process would cause issues.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

I fine-tuned SDXL on AscendNPU, and the results are good. I hope diffusers can support more devices.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

- Training examples: @sayakpaul

### CODE REVIEW

**Comment by HelloWorldBeginner** [comment]
I found some errors in checks, how can I fix it?   
```
examples/controlnet/train_controlnet_sdxl.py:16:1: I001 [*] Import block is un-sorted or un-formatted
examples/text_to_image/train_text_to_image_lora_sdxl.py:18:1: I001 [*] Import block is un-sorted or un-formatted
src/diffusers/models/activations.py:16:1: I001 [*] Import block is un-sorted or un-formatted
src/diffusers/models/attention_processor.py:14:1: I001 [*] Import block is un-sorted or un-formatted
```
It's strange because I didn't modify the code here.

**Comment by sayakpaul** [comment]
You can do the following:

* Create a fresh Python environment.
* Run `pip install -e ".[quality]"` from the root of `diffusers`.
* Run `make style && make quality`.

**Comment by HelloWorldBeginner** [comment]
I've already fixed the code formatting issues in the checks.

**Comment by yiyixuxu** [comment]
@sayakpaul 
I'm ok with this PR if you think it is needed :)

**Comment by sayakpaul** [comment]
Thanks, Yiyi. 

I am alright with the PR because the number of changes is extremely minimal.

File: src/diffusers/models/attention_processor.py
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:1291]: Hmm, so when Torch NPU is available it will default to using `torch_npu.npu_fusion_attention` right? But our current library-wide behaviour is that when PyTorch 1.x is available we rely on `AttnProcessor` and when PyTorch 2.x is available we rely on `AttnProcessor2_0` which uses `F.scaled_dot_product_attention()`. These are two default attention processors we use based on the available PyTorch version. 

So, with that in mind, I find this to be slightly problematic as we are moving away from a conceptual understanding. Folks that use the library already know that `F.scaled_dot_product_attention()` is being used if they're using PyTorch 2.x unless stated otherwise. Therefore, I think it might be better to have an `AttnProcessorNPU` class and use that instead. In that class, we will be able to do proper error handling too such as if `query.dtype` is not `torch.float16` or `torch.blfoat16`, we error out. 

I would like to ask @yiyixuxu and @DN6 for their opinions here too.

Comment (HelloWorldBeginner) [reply src/diffusers/models/attention_processor.py:1291]: Alright, I understand your point. It would be a good idea to separate the torch_npu flash attention module.

**Comment by HelloWorldBeginner** [comment]
I've separated the NPU flash attention into a module and implemented a switch control using parameters.
I've tested it and it works.

File: examples/text_to_image/train_text_to_image_lora_sdxl.py
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_lora_sdxl.py:634]: Have an attention processor mechanism. So it should look like: `unet.set_attn_processor(NPUAttnProcessor())`.

File: src/diffusers/models/activations.py
Comment (sayakpaul) [inline src/diffusers/models/activations.py:107]: Can we not use the existing `self.gelu()` when using NPU?

File: src/diffusers/models/attention_processor.py
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:230]: We don't need to do it like that, actually. IMO, `set_attn_processor(AttnProcessorNPU())` on the underlying UNet should be fine here.

**Review by sayakpaul — APPROVED** [review]
Thanks for working on this.

File: src/diffusers/models/activations.py
Comment (yiyixuxu) [inline src/diffusers/models/activations.py:110]:
```suggestion
          hidden_states = self.proj(hidden_states)
          if if is_torch_npu_available():
            return torch_npu.npu_geglu(hidden_states, dim=-1, approximate=1)[0]
        else:
            hidden_states, gate = hidden_states.chunk(2, dim=-1)
            return hidden_states * self.gelu(gate)
```

File: src/diffusers/models/activations.py
Comment (HelloWorldBeginner) [reply src/diffusers/models/activations.py:107]: Compared to self.gelu(), using torch_npu.npu_geglu can run faster and save memory on NPU.

**Comment by HelloWorldBeginner** [comment]
> Thanks for working on this.
> 
> For me, the following would be nice to add before we merge:
> 
> * Documentation -- add an entry about the NPU processor to https://huggingface.co/docs/diffusers/main/en/api/attnprocessor
> * Test: Similar to https://github.com/huggingface/diffusers/blob/26a7851e1e0b18da746d6ae80bb105050f7187e0/tests/models/test_modeling_common.py#L307
> 
> @yiyixuxu could you review the changes introduced the core modules of the library and comment?

Sure, I'll add unit tests and documentation later.

**Comment by HelloWorldBeginner** [comment]
I've updated the code. @sayakpaul

**Comment by HelloWorldBeginner** [comment]
Hi @sayakpaul.
I noticed the PR is still open. Does the code still need review from others?

**Comment by sayakpaul** [comment]
Yes, it needs reviews from our core maintainer @yiyixuxu/

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Pull request merged**
